plugins/wistron-dock/fu-wistron-dock-device.c

```diff
@@ -0,0 +1,944 @@
+{
+	guint8 cmd[8] = {FU_WISTRON_DOCK_ID_DOCK_IMG_CONTROL, FU_WISTRON_DOCK_CMD_DFU_ADDRESS};
+	fu_memwrite_uint32(cmd + 2, addr, G_BIG_ENDIAN);
+	return fu_wistron_dock_device_control_write(self, cmd, sizeof(cmd), TRUE, error);
+}
+
+static gboolean
+fu_wistron_dock_device_write_img_data(FuWistronDockDevice *self,
+				      const guint8 *buf,
+				      gsize bufsz,
+				      GError **error)
+{
+	guint8 cmd[8] = {FU_WISTRON_DOCK_ID_DOCK_IMG_CONTROL,
+			 FU_WISTRON_DOCK_CMD_DFU_WRITEIMG_BLOCK};
+	guint8 towrite[FU_WISTRON_DOCK_TRANSFER_BLOCK_SIZE + 1] = {
+	    FU_WISTRON_DOCK_ID_DOCK_IMG_DATA};
+	if (!fu_memcpy_safe(towrite,
+			    sizeof(towrite),
+			    0x1, /* dst */
+			    buf,
+			    bufsz,
+			    0x0, /* src */
+			    bufsz,
+			    error))
+		return FALSE;
+	return fu_wistron_dock_device_data_write(self,
+						 cmd,
+						 sizeof(cmd),
+						 towrite,
+						 sizeof(towrite),
+						 error);
+}
+
+static gboolean
+fu_wistron_dock_device_write_blocks(FuWistronDockDevice *self,
+				    GPtrArray *chunks,
+				    FuProgress *progress,
+				    GError **error)
+{
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+
+		/* set address */
+		if (!fu_wistron_dock_device_set_img_address(self,
+							    fu_chunk_get_address(chk),
+							    error)) {
+			g_prefix_error(error,
+				       "failed to set img address 0x%x",
+				       fu_chunk_get_address(chk));
+			return FALSE;
+		}
+
+		/* write */
+		if (!fu_wistron_dock_device_write_img_data(self,
+							   fu_chunk_get_data(chk),
+							   fu_chunk_get_data_sz(chk),
+							   error)) {
+			g_prefix_error(error,
+				       "failed to write img data 0x%x",
+				       fu_chunk_get_address(chk));
+			return FALSE;
+		}
+
+		/* update progress */
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_wistron_dock_device_prepare_firmware(FuDevice *device,
+					GBytes *fw,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_archive_firmware_new();
+	g_autoptr(FuFirmware) fw_cbin = NULL;
+	g_autoptr(FuFirmware) fw_new = fu_firmware_new();
+	g_autoptr(FuFirmware) fw_wdfl = NULL;
+	g_autoptr(FuFirmware) fw_wdfl_sig = NULL;
+	g_autoptr(GPtrArray) imgs = NULL;
+
+	/* unzip */
+	if (!fu_firmware_parse(firmware, fw, flags, error))
+		return NULL;
+
+	imgs = fu_firmware_get_images(firmware);
+	for (guint i = 0; i < imgs->len; i++) {
+		FuFirmware *img = g_ptr_array_index(imgs, i);
+		const gchar *fn = fu_firmware_get_id(img);
+		if (g_str_has_suffix(fn, ".wdfl.sig")) {
+			if (fw_wdfl_sig != NULL) {
+				g_set_error_literal(error,
+						    G_IO_ERROR,
+						    G_IO_ERROR_INVALID_DATA,
+						    "multiple WDFL signatures found in archive");
+				return NULL;
+			}
+			if (fu_firmware_get_size(img) != FU_WISTRON_DOCK_WDFL_SIG_SIZE) {
```

Comment:
As a future proofing - perhaps this shouldn't be checked.  It would allow newer (longer) detached signature types in the future.

Reply:
Good idea. I've changed it to `<` as any newer version isn't going to be any smaller.